src/web/components/StickyBottomBanner/bannerPicker.ts

```diff
@@ -0,0 +1,102 @@
+import { record } from '@root/src/web/browser/ophan/ophan';
+
+export type MaybeFC = React.FC | null;
+type ShowBanner = (meta?: any) => MaybeFC;
+
+export type CanShowResult = {
+    result: boolean;
+    meta?: any;
+};
+
+type Banner = {
+    id: string;
+    show: ShowBanner;
+    canShow: () => Promise<CanShowResult>;
+    timeoutMillis?: number | null;
+};
+
+type BannerWithTimeout = Banner & {
+    cancelTimeout: () => void;
+};
+
+export type BannerConfig = Banner[];
+
+const DEFAULT_BANNER_TIMEOUT_MILLIS = 2000;
+
+const recordBannerTimeoutInOphan = (bannerId: string) =>
+    record({
+        component: 'banner-picker-timeout-dcr',
+        value: bannerId,
+    });
+
+const timeoutify = (banner: Banner): BannerWithTimeout => {
+    let timer: number | undefined;
+
+    const canShow = (): Promise<CanShowResult> =>
+        new Promise((resolve) => {
+            const timeout =
+                typeof banner.timeoutMillis === 'undefined'
```

Comment:
Having such different meanings for null and undefined worries me a bit, is this a common pattern?
I think I'd prefer making `timeoutMillis` a mandatory field:
`timeoutMillis: number | null`

Reply:
Yeah that's a really good point, I can see that this is a little confusing now. My intention was around the difference between omitting the property (which gives you the default 2s timeout):

```js
{
  id: 'some-banner',
  // etc
}
```

and including it but specifying `null` (which gives no timeout):

```js
{
  id: 'some-banner',
  // etc
  timeoutMillis: null,
}
```

But your suggestion of making it a mandatory field sounds sensible. I think the reasoning behind having a default was that when adding a new banner you wouldn't have to think about the timeout too much. But realistically new banners aren't added that often, and you probable _do_ need to think about the timeout anyway. Perhaps making it mandatory but defining a const `DEFAULT_TIMEOUT = 2000` would keep the intent clear.

Reply:
Makes sense - the developer has to consider whether or not they want a timeout

Reply:
I've made `timeoutMillis` mandatory in this commit: d4b8e269aa729d633d9bc43177ccc5b9182cf21a.